File: Python3.x/75-Sort Colors.py
# call lib
class Solution:
    def sortColors(self, nums: 'List[int]') -> 'None':
        """
        Do not return anything, modify nums in-place instead.
        """
        nums.sort()

# quick sort
class Solution:
    def sortColors(self, nums: 'List[int]') -> 'None':
        """
        Do not return anything, modify nums in-place instead.
        """
        def sort_recursion(array, left, right):
            if left >= right: return  # 递归停止
            i, j = left, right
            flag = array[left]  # 取首位为对比标志值
            while i < j:  # 找flag的最终位置
                while i < j and array[j] >= flag:  # 用j向左扫描找小于flag的记录
                    j -= 1
                if i < j:
                    array[i] = array[j]  # 小记录移到左边
                    i += 1
                while i < j and array[i] <= flag:  # 用i向右扫描找大于flag的记录
                    i += 1
                if i < j:
                    array[j] = array[i]  # 大记录移到右边
                    j -= 1
            array[i] = flag  # 将对比的标志位存入其最终位置
            sort_recursion(array, left, i - 1)  # 左边
            sort_recursion(array, i + 1, right)  # 右边

        sort_recursion(nums, 0, len(nums) - 1)

File: Algorithms/C++/75-Sort_Colors.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        sort(nums.begin(), nums.end());
    }
};


// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Sort Colors.
// Memory Usage: 8.6 MB, less than 87.72% of C++ online submissions for Sort Colors.
class Solution {
public:
    void sortColors(vector<int>& nums) {
        quick_sort(nums, 0, nums.size() - 1);
    }
private:
    void quick_sort(vector<int>& nums, int l, int r) {
        if (l >= r)
            return;
        int left = l, right = r;
        int flag = nums[l];
        while (left < right) {
            while (left < right && nums[right] >= flag)
                --right;
            if (left < right) {
                nums[left] = nums[right];
                ++left;
            }
            while (left < right && nums[left] <= flag)
                ++left;
            if (left < right) {
                nums[right] = nums[left];
                --right;
            }
            nums[left] = flag;
            quick_sort(nums, l, left - 1);
            quick_sort(nums, left + 1, r);
        }
    }
};
